### PR TITLE
Access Rules update

### PR DESCRIPTION
**Pull request by heylemur**
Summary of changes:
- Removed the granting access examples from the Managing_Accounts folders
- Deleted the Composite_Roles examples
- Added tags to the Managing_Resources examples
- Replaced the Managing_Roles examples with the new
- Added info on migrating to access rules to the Readme files

### CODE REVIEW

**Comment by turner-strongdm** [comment]
Everything in this one looks good.

**Review by turner-strongdm — APPROVED** [review]
Didn't submit as review. My bad.

Everything looks good from my perspective.

File: 3_managing_roles/README.md [hunk 1]
```diff
@@ -1,3 +1,108 @@
...
 
-A collection of examples regarding managing roles.
+A collection of examples regarding managing Roles.
+
+## Migrating to Access Rules
+
+To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type. Role Grants will be removed from the Terraform
+provider by March 15, 2022.
```
Comment (jeffreylees) [inline 3_managing_roles/README.md:10]: Before we publish we should reaffirm this date. Or better yet, if it's really that soon, not mention it. I'll ask about this.

File: 3_managing_roles/README.md [hunk 2]
```diff
@@ -1,3 +1,108 @@
...
+The following examples demonstrate the deprecated Role Grants, Dynamic Access
+Rules with Tags and Resource Types, and Static Access Rules for backwards
+compatibility with Role Grants.
+
+### Role Grants (deprecated)
+
+Previously, you would grant a Role access to specific resources by ID via Role
+Grants:
+
+```hcl
```
Comment (jeffreylees) [inline 3_managing_roles/README.md:21]: `hcl` is more generic than `tf` or `terraform` - is there a reason for that here, do you know? Local previews suggest that our syntax highlighter takes all three.

File: 3_managing_roles/README.md [hunk 3]
```diff
@@ -1,3 +1,108 @@
...
+
+resource "sdm_role_grant" "engineering-postgres" {
+  resource_id = sdm_resource.postgres-test.id
+  role_id = sdm_role.engineering.id
+}
+```
+
+### Dynamic Access Rules
+
+When using Access Rules, the best practice is to grant Resources access based on
```
Comment (jeffreylees) [inline 3_managing_roles/README.md:66]: This sounds like we are giving Resources some kind of access. Maybe we could reword to make it clear that we're giving Roles access to Resources based on type and tags.

File: README.md [hunk 4]
```diff
@@ -2,7 +2,13 @@
...
-> **NOTE:**
-> This repository is currently under construction. The content is expected to change in the coming weeks.
-
+> **NOTE:**  
+> To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type.
+>
+> Previously, you would grant a Role access to specific resources by ID via Role
+Grants. Now, when using Access Rules, the best practice is to grant Resources access based on Type and Tags.
```
Comment (jeffreylees) [inline README.md:11]: Same as above

File: README.md [hunk 5]
```diff
@@ -2,7 +2,13 @@
 This is the examples repository for the [strongDM Terraform Provider](https://github.com/strongdm/terraform-provider-sdm).
 
 ---
-> **NOTE:**
-> This repository is currently under construction. The content is expected to change in the coming weeks.
-
+> **NOTE:**  
+> To increase flexibility when managing thousands of Resources, Role Grants have
```
Comment (jeffreylees) [inline README.md:6]: I'd maybe just say large numbers of resources? Surely Access Rules are more flexible with even mediocre numbers, I don't think you have to have thousands to realize the benefits.

**Review by jeffreylees — COMMENTED** [review]
Just a couple of suggestions/questions. Once those are resolved we should hold onto this one until the others are ready and V2 is live.

File: 3_managing_roles/README.md [hunk 6]
```diff
@@ -1,3 +1,108 @@
 # Managing Roles
 
-A collection of examples regarding managing roles.
+A collection of examples regarding managing Roles.
+
+## Migrating to Access Rules
+
+To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type. Role Grants will be removed from the Terraform
```
Comment (heylemur) [inline 3_managing_roles/README.md:9]:
```suggestion
on Resource Tags and Type.

[Role Grants will be removed from the Terraform]: #
```

File: 3_managing_roles/README.md (hunk 1 above)
Comment (heylemur) [inline 3_managing_roles/README.md:10]:
```suggestion
```

File: 3_managing_roles/README.md (hunk 2 above)
Comment (heylemur) [reply 3_managing_roles/README.md:21]: Oh, `hcl` is what was in the migration notes from Product, and I left in in there. I can change to `tf`.

Comment (heylemur) [inline 3_managing_roles/README.md:21]:
```suggestion
```tf
```

File: 3_managing_roles/README.md [hunk 7]
```diff
@@ -1,3 +1,108 @@
...
+  role_id = sdm_role.engineering.id
+}
+```
+
+### Dynamic Access Rules
+
+When using Access Rules, the best practice is to grant Resources access based on
+Type and Tags.
+
+```hcl
```
Comment (heylemur) [inline 3_managing_roles/README.md:69]:
```suggestion
```tf
```

File: 3_managing_roles/grant_access.tf [hunk 8]
```diff
@@ -14,39 +14,73 @@
...
 }
 
 #################
-# Grant the role access to the datasource
+# Grant access
 #################
-resource "sdm_role_grant" "example_role_grant" {
-  role_id     = sdm_role.example_user.id
-  resource_id = sdm_resource.postgres_example.id
+# When using Dynamic Access Rules, the best practice is to grant Resources access based on
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:44]:
```suggestion
# When using Access Rules, the best practice is to give Roles access to Resources based on
```

File: 3_managing_roles/grant_access.tf [hunk 9]
```diff
@@ -14,39 +14,73 @@
...
+Type and Tags.
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:45]:
```suggestion
type and tags.
```

File: README.md (hunk 5 above)
Comment (heylemur) [inline README.md:6]:
```suggestion
> To increase flexibility when managing a large volume of Resources, Role Grants have
```

File: README.md (hunk 4 above)
Comment (heylemur) [inline README.md:11]:
```suggestion
Grants. Now, when using Access Rules, the best practice is to give Roles access to Resources based on type and tags.
```

File: README.md [hunk 10]
```diff
@@ -2,7 +2,13 @@
...
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type.
+>
+> Previously, you would grant a Role access to specific resources by ID via Role
+Grants. Now, when using Access Rules, the best practice is to grant Resources access based on Type and Tags.
+>
+>The following examples demonstrate Dynamic Access Rules with Tags and Resource Types, as well as Static Access Rules for backwards compatibility with Role Grants. If it is _necessary_ to grant access to specific Resources in the same way as Role Grants did, you can use Resource IDs directly in Access Rules.
```
Comment (heylemur) [inline README.md:13]:
```suggestion
>The following examples demonstrate Dynamic Access Rules with tags and Resource types, as well as Static Access Rules for backwards compatibility with Role Grants. If it is _necessary_ to grant access to specific Resources in the same way as Role Grants did, you can use Resource IDs directly in Access Rules.
```

File: 3_managing_roles/README.md [hunk 11]
```diff
@@ -1,3 +1,108 @@
...
+
+## Migrating to Access Rules
+
+To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type. Role Grants will be removed from the Terraform
+provider by March 15, 2022.
+
+The following examples demonstrate the deprecated Role Grants, Dynamic Access
+Rules with Tags and Resource Types, and Static Access Rules for backwards
```
Comment (heylemur) [inline 3_managing_roles/README.md:13]:
```suggestion
Rules with tags and Resource types, and Static Access Rules for backwards
```

File: README.md [hunk 12]
```diff
@@ -2,7 +2,13 @@
 This is the examples repository for the [strongDM Terraform Provider](https://github.com/strongdm/terraform-provider-sdm).
 
 ---
-> **NOTE:**
-> This repository is currently under construction. The content is expected to change in the coming weeks.
-
+> **NOTE:**  
+> To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type.
```
Comment (heylemur) [inline README.md:8]:
```suggestion
on Resource tags and type.
```

File: 3_managing_roles/README.md [hunk 13]
```diff
@@ -1,3 +1,108 @@
...
+compatibility with Role Grants.
+
+### Role Grants (deprecated)
+
+Previously, you would grant a Role access to specific resources by ID via Role
```
Comment (heylemur) [inline 3_managing_roles/README.md:18]:
```suggestion
Previously, you would grant a Role access to specific Resources by ID via Role
```

File: README.md [hunk 14]
```diff
@@ -2,7 +2,13 @@
...
+>
+> Previously, you would grant a Role access to specific resources by ID via Role
```
Comment (heylemur) [inline README.md:10]:
```suggestion
> Previously, you would grant a Role access to specific Resources by ID via Role
```

File: 3_managing_roles/README.md [hunk 15]
```diff
@@ -1,3 +1,108 @@
...
+## Migrating to Access Rules
+
+To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type. Role Grants will be removed from the Terraform
+provider by March 15, 2022.
+
+The following examples demonstrate the deprecated Role Grants, Dynamic Access
+Rules with Tags and Resource Types, and Static Access Rules for backwards
+compatibility with Role Grants.
```
Comment (heylemur) [inline 3_managing_roles/README.md:14]:
```suggestion
```

File: 3_managing_roles/README.md [hunk 16]
```diff
@@ -1,3 +1,108 @@
...
+  }
+}
+```
+
+### Static Access Rules
+
+If it is _necessary_ to grant access to specific Resources in the same way as
+Role Grants did, you can use Resource IDs directly in Access Rules.
+
+```hcl
```
Comment (heylemur) [inline 3_managing_roles/README.md:101]:
```suggestion
```tf
```

File: 3_managing_roles/grant_access.tf [hunk 17]
```diff
@@ -14,39 +14,73 @@
 #
 
 #################
-# Create a datasource
+# Create a resource (e.g., Postgres)
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:17]:
```suggestion
# Create a Resource (e.g., Postgres)
```

File: 3_managing_roles/grant_access.tf [hunk 18]
```diff
@@ -14,39 +14,73 @@
...
-resource "sdm_role_grant" "example_role_grant" {
-  role_id     = sdm_role.example_user.id
-  resource_id = sdm_resource.postgres_example.id
+# When using Dynamic Access Rules, the best practice is to grant Resources access based on
+Type and Tags.
+
+resource "sdm_role" "example-role" {
+  name = "example-role"
+
+  # Example: Grant access to all dev environment resources in us-west region
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:50]:
```suggestion
  # Example: Grant access to all dev environment Resources in us-west region
```

File: 3_managing_roles/grant_access.tf [hunk 19]
```diff
@@ -14,39 +14,73 @@
...
+  access_rule {
+    tags = {
+      env = "dev"
+      region = "us-west"
+    }
+  }
+
+  # Example: Grant access to all Postgres resources
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:58]:
```suggestion
  # Example: Grant access to all Postgres Resources
```

File: 3_managing_roles/grant_access.tf [hunk 20]
```diff
@@ -14,39 +14,73 @@
...
+  # Grant access to all Redis Datasources in us-east region
+  access_rule {
+    type = "redis"
+    tags = {
+      region = "us-east"
+    }
+  }
+}
+
+# If it is _necessary_ to grant access to specific resources in the same way as
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:72]:
```suggestion
# If it is _necessary_ to grant access to specific Resources in the same way as
```

File: 3_managing_roles/grant_access.tf [hunk 21]
```diff
@@ -14,39 +14,73 @@
...
+RoleGrants did, you can use Resource IDs directly in Static Access Rules.
```
Comment (heylemur) [inline 3_managing_roles/grant_access.tf:73]:
```suggestion
Role Grants did, you can use Resource IDs directly in Static Access Rules.
```

File: 3_managing_roles/README.md [hunk 22]
```diff
@@ -1,3 +1,108 @@
 # Managing Roles
 
-A collection of examples regarding managing roles.
+A collection of examples regarding managing Roles.
+
+## Migrating to Access Rules
+
+To increase flexibility when managing thousands of Resources, Role Grants have
```
Comment (heylemur) [inline 3_managing_roles/README.md:7]:
```suggestion
To increase flexibility when managing a large volume of Resources, Role Grants have
```

File: 3_managing_roles/README.md [hunk 23]
```diff
@@ -1,3 +1,108 @@
...
+### Dynamic Access Rules
+
+When using Access Rules, the best practice is to grant Resources access based on
+Type and Tags.
+
+```hcl
+resource "sdm_role" "engineering" {
+  name = "engineering"
+
+  # Grant access to all dev environment resources in us-west
```
Comment (heylemur) [inline 3_managing_roles/README.md:73]:
```suggestion
  # Grant access to all dev environment Resources in us-west
```

File: 3_managing_roles/README.md [hunk 24]
```diff
@@ -1,3 +1,108 @@
...
+  access_rule {
+    tags = {
+      env = "dev"
+      region = "us-west"
+    }
+  }
+
+  # Grant access to all postgres resources
```
Comment (heylemur) [inline 3_managing_roles/README.md:81]:
```suggestion
  # Grant access to all Postgres Resources
```

File: 3_managing_roles/README.md [hunk 25]
```diff
@@ -1,3 +1,108 @@
...
+  access_rule {
+    type = "postgres"
+  }
+
+  # Grant access to all redis datasources in us-east
```
Comment (heylemur) [inline 3_managing_roles/README.md:86]:
```suggestion
  # Grant access to all Redis Datasources in us-east
```

**Review by jeffreylees — APPROVED** [review]

**Comment by etodd** [comment]
Bad news - we had to change the shape of the Terraform provider _again..._ this time back to its original shape: a JSON string.

So instead of
```
sdm_role "blah" {
    access_rule {
    ...
    }
    access_rule {
    ...
    }
}
```
It's back to
```
sdm_role "blah" {
    access_rules = "[ {...}, {...} ]"
}
```

I'm so sorry. 🙏

**Comment by heylemur** [comment]
@etodd No worries--thank you for telling us! I'll update the examples.

**Comment by turner-strongdm** [comment]
Ok since you're swapping to the `jsonencode` it's expecting it to be` json`. Using the `=` rather than `:` could error. I have some examples I pulled from our docs that I had been sending to customers. 


Static Role:

```HCL
resource "sdm_role" "people" {
  name = "People"
  access_rules = jsonencode(
  [{
    "ids": [
        "rs-0c3b76e560f5f37b",
        "rs-22904ced60ac13ec",
        "rs-25b92097609bed25",
        "rs-2d396d0260a810cb",
        "rs-2f1d8408609bf95a",
        "rs-74b6183d60f5e06b"
    ]
  }]
)
}
```

Single Dynamic Role:

```HCL
resource "sdm_role" "people" {
  name = "People"
  access_rules = jsonencode(
  [
    {
      "type": "redis"
      "tags": { "foo": "bar" }
    }
  ]
)
}
```

Multi Rule Dynamic Role:

```HCL
resource "sdm_role" "people" {
  name = "People"
  access_rules = jsonencode(
  [
    {
      "ids": [ "${sdm_resource.ds1.id}" ]
    },
    {
      "type": "redis"
    },
    {
      "type": "redis"
      "tags": { "foo": "bar" }
    }
  ]
)
}
```

**Comment by philhassey** [comment]
@turner-strongdm Can

` [ "${sdm_resource.ds1.id}" ]`

Be

`[sdm_resource.ds1.id]`

instead?  That would look so much better.  The whole point of switching to jsonencode is to avoid string interpolation.

**Comment by turner-strongdm** [comment]
Yes. Sorry I was copying. Either is acceptable. The second is only post `v.12` or `.13` so if they are running older version may not work, but we do have a requirement for higher version in our provider so wouldn't be a bad to keep to newer format.

**Comment by etodd** [comment]
I did some testing and it looks like Terraform supports both `=` and `:`:

```
resource "sdm_role" "role_example" {
  name = "Example Role 2"
  access_rules = jsonencode([
    {
      ids = [sdm_resource.postgres_example.id]
    },
    {
      "type": "redis"
    }
  ])
}
```

The second way (`"field": "value"`) is closer to the underlying JSON representation, so it's probably better to do it that way.

**Comment by heylemur** [comment]
@etodd and @turner-strongdm Thank you for the examples and info! I changed the equal signs to colons and fixed the formatting. Does this look OK?

File: 3_managing_roles/README.md [hunk 26]
```diff
@@ -69,25 +69,28 @@ resource "sdm_role" "engineering" {
   name = "engineering"
 
   # Grant access to all dev environment Resources in us-west
-  access_rules = jsonencode(
-    [
-      { tags = { env = "dev", region = "us-west" } }
-    ])
+  access_rules = jsonencode([
+    {
+      "tags": { "env": "dev" }, { "region": "us-west" }
```
Comment (etodd) [inline 3_managing_roles/README.md:74]: This should be:
```
"tags": { "env": "dev", "region": "us-west" }
```

File: 1_managing_resources/create_eks_cluster.tf [hunk 27]
```diff
@@ -14,22 +14,21 @@
...
-        certificate_authority          = local.certificate_authority,
-        certificate_authority_filename = "random_string",
-        access_key                     = "AKIAIOSFODNN7EXAMPLE",
-        secret_access_key              = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
-        role_arn                       = "arn:aws:iam::000000000000:role/RoleName",
-        healthcheck_namespace          = "default",
-        tags                           = { region = "us-east", env = "dev" }
-      }
-    ])
+  amazon_eks = jsonencode([
```
Comment (etodd) [inline 1_managing_resources/create_eks_cluster.tf:17]: No jsonencode here

File: 1_managing_resources/create_postgres_datasource.tf [hunk 28]
```diff
@@ -14,16 +14,15 @@
...
-        name     = "Example Postgres Datasource",
-        hostname = "example.strongdm.com",
-        database = "example",
-        username = "example",
-        password = "example",
-        port     = 5432,
-        tags     = { env = "dev" }
-      }
-    ]) 
+  postgres = jsonencode([
```
Comment (etodd) [inline 1_managing_resources/create_postgres_datasource.tf:17]: No jsonencode here

File: 1_managing_resources/create_rdp_server.tf [hunk 29]
```diff
@@ -14,15 +14,14 @@
...
-      {
-        name     = "Example RDP Server",
-        hostname = "example.strongdm.com",
-        username = "example",
-        password = "example",
-        port     = 3389,
-        tags     = { env = "dev" }
-      }
-    ])
+  rdp = jsonencode([
```
Comment (etodd) [inline 1_managing_resources/create_rdp_server.tf:17]: No jsonencode here

File: 1_managing_resources/create_redis_datasource.tf [hunk 30]
```diff
@@ -14,14 +14,13 @@
...
-    [
-      {
-        name     = "redis example",
-        hostname = "example.com",
-        password = "example",
-        port     = 6379,
-        tags = { env = "dev" }
-      }
-    ])
+  redis = jsonencode([
```
Comment (etodd) [inline 1_managing_resources/create_redis_datasource.tf:17]: No jsonencode here

File: 1_managing_resources/create_ssh_server.tf [hunk 31]
```diff
@@ -14,14 +14,13 @@
...
-    [
-      {
-        name     = "Example SSH Server",
-        hostname = "example.strongdm.com",
-        username = "example",
-        port     = 22,
-        tags     = { env = "dev" }
-      }
-    ])
+  ssh = jsonencode([
```
Comment (etodd) [inline 1_managing_resources/create_ssh_server.tf:17]: No jsonencode here

File: 2_managing_accounts/account_user.tf [hunk 32]
```diff
@@ -14,19 +14,18 @@
...
-  user = jsonencode(
-    [
-      {
-        first_name = "Alice",
-        last_name  = "Belcher",
-        email      = "alice.belcher@strongdm.com",
-        suspended  = false
-      }
-    ])
+  user = jsonencode([
```
Comment (etodd) [inline 2_managing_accounts/account_user.tf:17]: No jsonencode here

File: 2_managing_accounts/assign_accounts_to_roles.tf [hunk 33]
```diff
@@ -17,32 +17,33 @@
...
-  user = jsonencode(
-    [
-      {
-        first_name = "Example",
-        last_name  = "Example",
-        email      = "example@strongdm.com",
-        suspended  = false
-      }
-    ])
+  user = jsonencode([
```
Comment (etodd) [inline 2_managing_accounts/assign_accounts_to_roles.tf:20]: No jsonencode here

File: 2_managing_accounts/create_service_account.tf [hunk 34]
```diff
@@ -13,16 +13,15 @@
...
+############################
 resource "sdm_account" "example_service_account" {
-  service = jsonencode(
-    [
-      {
-        name      = "Example Service Account",
-        suspended = false
-      }
-    ])
+  service = jsonencode([
```
Comment (etodd) [inline 2_managing_accounts/create_service_account.tf:20]: No jsonencode here

File: 2_managing_accounts/create_user.tf [hunk 35]
```diff
@@ -17,14 +17,13 @@
...
-  user = jsonencode(
-    [
-      {
-        first_name = "Example",
-        last_name  = "Example",
-        email      = "example@strongdm.com",
-        suspended  = false
-      }
-    ])
+  user = jsonencode([
```
Comment (etodd) [inline 2_managing_accounts/create_user.tf:20]: No jsonencode here

File: 3_managing_roles/grant_access.tf [hunk 36]
```diff
@@ -13,96 +13,103 @@
...
-        name                       = "Example Postgres Resource",
-        hostname                   = "example.strongdm.com",
-        port                       = "1306",
-        database                   = "postgres",
-        secret_store_id            = "se-123e45678901bb23",
-        secret_store_username_path = "example/sdm?key=user",
-        secret_store_password_path = "example/sdm?key=pw"
-      }
-    ])
+  postgres = jsonencode([
```
Comment (etodd) [inline 3_managing_roles/grant_access.tf:20]: No jsonencode here

File: 3_managing_roles/grant_access.tf [hunk 37]
```diff
@@ -13,96 +13,103 @@
...
 
   # Example: Grant access to all dev environment Resources in us-west region
-  access_rules = jsonencode(
-    [
-      { tags = { env = "dev", region = "us-west" } }
-    ])
-  }
+  access_rules = jsonencode([
+    {
+      "tags": {"env": "dev" }, { "region": "us-west" }
```
Comment (etodd) [inline 3_managing_roles/grant_access.tf:58]: This should be:
```
"tags": { "env": "dev", "region": "us-west" }
```

File: 3_managing_roles/grant_access.tf [hunk 38]
```diff
@@ -13,96 +13,103 @@
...
-  user = jsonencode(
-    [
-      {
-        first_name = "example",
-        last_name  = "example",
-        email      = "example@strongdm.com",
-        suspended  = false
-      }
-    ])
+  user = jsonencode([
```
Comment (etodd) [inline 3_managing_roles/grant_access.tf:99]: No jsonencode here

File: 4_managing_gateways/create_gateway.tf [hunk 39]
```diff
@@ -14,11 +14,10 @@
...
 
 resource "sdm_node" "example_gateway" {
-  gateway = jsonencode(
-    [
-      {
-        name           = "example-gateway",
-        listen_address = "gateway.example.com:5555"
-      }
-    ])
+  gateway = jsonencode([
```
Comment (etodd) [inline 4_managing_gateways/create_gateway.tf:17]: No jsonencode here

File: 4_managing_gateways/create_relay.tf [hunk 40]
```diff
@@ -14,8 +14,9 @@
 #
 
 resource "sdm_node" "example_relay" {
-  relay = jsonencode(
-    [
-      { name = "example-relay" }
-    ])
+  relay = jsonencode([
```
Comment (etodd) [inline 4_managing_gateways/create_relay.tf:17]: No jsonencode here

File: 1_managing_resources/create_redis_datasource.tf (hunk 30 above)
Comment (etodd) [reply 1_managing_resources/create_redis_datasource.tf:17]: This should be:
```
resource "sdm_resource" "redis_example" {
  redis {
    name     = "redis example"
    hostname = "example.com"
    password = "example"
    port     = 6379
    tags = { env = "dev" }
  }
}
```

**Review by etodd — CHANGES_REQUESTED** [review]
We only want to change the `role.access_rules` field to be JSON. All the other stuff (resources, accounts, account attachments, etc.) should stay the way they were. I put comments on the places that should be reverted back, and provided one example of the way it should look once it's reverted.

**Comment by heylemur** [comment]
Thank you @etodd for your comments and examples! It's fixed now.

File: 3_managing_roles/README.md [hunk 41]
```diff
@@ -1,3 +1,114 @@
...
+}
+```
+
+### Dynamic Access Rules
+
+When using Access Rules, the best practice is to give Roles access to Resources based on
+type and tags.
+
+```tf
+resource "sdm_role" "engineering" {
```
Comment (etodd) [inline 3_managing_roles/README.md:68]: this block should be:

```
resource "sdm_role" "engineering" {
  name = "engineering"
  access_rules = jsonencode([

    # Grant access to all dev environment Resources in us-west
    {
      "tags": { "env": "dev", "region": "us-west" }
    },

    # Grant access to all Postgres Resources
    {
      "type": "postgres"
    },

    # Grant access to all Redis Datasources in us-east
    {
      "type": "redis",
      "tags": { "region": "us-east" }
    }
  ])
}
```

File: 3_managing_roles/grant_access.tf [hunk 42]
```diff
@@ -13,54 +13,99 @@
...
-    password = "example"
+################
+# Grant access
+################
+# When using Access Rules, the best practice is to give Roles access to Resources based on
+type and tags.
 
-    database = "example"
-  }
+resource "sdm_role" "example-role" {
```
Comment (etodd) [inline 3_managing_roles/grant_access.tf:50]: this block should be:

```
resource "sdm_role" "example-role" {
  name = "example-role"
  access_rules = jsonencode([

    # Example: Grant access to all dev environment Resources in us-west region
    {
      "tags": { "env": "dev", "region": "us-west" }
    },

    # Example: Grant access to all Postgres Resources
    {
      "type": "postgres"
    },

    # Grant access to all Redis Datasources in us-east region
    {
      "type": "redis",
      "tags": { "region": "us-east" }
    }
  ])
}
```

**Review by etodd — CHANGES_REQUESTED** [review]
Just a couple other changes!

File: 3_managing_roles/README.md (hunk 41 above)
Comment (heylemur) [reply 3_managing_roles/README.md:68]: Thank you!

File: 3_managing_roles/grant_access.tf (hunk 42 above)
Comment (heylemur) [reply 3_managing_roles/grant_access.tf:50]: Thanks!

**Review by etodd — APPROVED** [review]

**Pull request merged**